# 1 Import Libraries and Files

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Import path and csv files
path= r'/Users/owner/Documents/05_2023_Instacart_Basket_Analysis'

In [10]:
df_dept = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))

In [11]:
df_dept.shape

(21, 2)

In [12]:
df_dept.head(21)

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [3]:
#import orders_products_merged.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [6]:
ords_prods_merge.shape

(32434212, 19)

In [7]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

          Busiest_days     Busiest_day Busiest_period_of_day  max_order  \
0  Regularly busy days  Regularly busy        Average Orders         10   
1      Least busy days  Regularly busy        Average Orders         10   
2      Least busy days  Regularly busy           Most Orders         10   
3      Least busy days      Least busy        Average Orders         10   
4      Least busy days      Least busy           Most Orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

# 2 Aggregate mean of order_number by department_id

In [9]:
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3 Analyze Results

Results from the entire dataframe are similar to subset, but it created an order_number mean for the whole dataframe. The minimum mean in the subset was 11.3 (department-babies) and the smallest mean in this datagrame is 15.2 (department-international). The biggest mean int he subset was 19.4 (department-household) while the largest mean in this subset it 20.2 (department-personal care). (The 21 department is the 'missing' department) 

In [16]:
#Drop columns and create new dataframe to complete this task
df_ords_prods_merge = ords_prods_merge.drop(['max_order', 'loyalty_flag'], axis = 1)

In [17]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

          Busiest_days     Busiest_day Busiest_period_of_day  
0  Regularly busy days  Regularly busy        Average Orders  
1      Least busy days  Regularly busy        Average Orders  
2      Least busy days  Regularly busy           Most Orders  
3      Least busy days      Least busy        Average Orders  
4      Least busy days      Least busy           Most Orders

# 4 Create Loyalty Flag 

In [18]:
df_ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [20]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

          Busiest_days     Busiest_day Busiest_period_of_day  max_order  
0  Regularly busy days  Regularly busy        Average Orders         10  
1      Least busy days  Regularly busy        Average Orders         10  
2      Least busy days  Regularly busy           Most Orders         10  
3      Least busy days      Least busy        Average Orders         10  
4      Least busy days      Least busy           Most Orders         10  
5  Regularly busy days  Regularly busy        Average Orders         10  
6         Busiest days  Regularly busy           Most Orders         10  
7         Busiest days  Regularly busy           Most Orders         10  
8         Busiest days  Regularly busy           Most Orders         10  
9      Least busy days      Least busy        Average Orders         10

In [21]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [22]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [23]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [25]:
df_ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices        price_label  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda      

In [27]:
# Check frequencies of loyalty flags 
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

# 5. mean prices of products for each type of customer

In [28]:
# Group by "prices" 
df_ords_prods_merge.groupby('prices')

In [29]:
# Split  data into groups by "loyalty_flag" and  calculate the mean for each "prices" column
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.388747
New customer      13.294370
Regular customer  12.496203

# 6. Create Spending Flag based on avg price across all orders

In [30]:
# Group data by "user_id" column
#generate mean product price for each user
#create "mean_price" for aggregation results
df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [31]:
df_ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices        price_label  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda      

In [32]:
# Create flags to assign "spending" labels to user ID based 
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [34]:
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [35]:
#Check frequency
df_ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31798751
High spender      635461
Name: spending_flag, dtype: int64

# 7 Create order frequency flag based on days_since_prior_order


In [10]:
# Group data by "user_id" column
#generate median days for each user
#create "median_days" for aggregation results
df_ords_prods_merge['median_days'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

NameError: name 'df_ords_prods_merge' is not defined

In [37]:
#Check Group
df_ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  prices        price_label         Busiest_days  \
0          Soda  ...     9.0  Mid-range product  Regularly busy days   
1          Soda  ...     9.0  Mid-range product      Least busy days   
2          Soda  ...     9.0  Mid-range product      Least busy days   
3          Soda  ...     9.0  Mid-range product      Least busy days   
4          Soda  ...     9.0  Mid-range product      Least busy days   
5          Soda  ...     9.0  Mid-range product  Regularly busy days   
6          Soda  ...     9.0  Mid-range product         Busiest days   
7          Soda  ...     9.0  Mid-range product         Busiest days   
8          Soda  ...     9.0  Mid-range product         Busiest days   
9          Soda  ...     9.0  Mid-range product      Least busy days   
10         Soda  ...     9.0  Mid-range product         Busiest days   
11       

In [38]:
#Create flags 
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [39]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days'] > 10) & (df_ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [40]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [41]:
df_ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21577409
Regular customer          7217134
Non-frequent customer     3639669
Name: frequency_flag, dtype: int64

In [42]:
df_ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...        price_label         Busiest_days     Busiest_day  \
0          Soda  ...  Mid-range product  Regularly busy days  Regularly busy   
1          Soda  ...  Mid-range product      Least busy days  Regularly busy   
2          Soda  ...  Mid-range product      Least busy days  Regularly busy   
3          Soda  ...  Mid-range product      Least busy days      Least busy   
4          Soda  ...  Mid-range product      Least busy days      Least busy   
5          Soda  ...  Mid-range product  Regularly busy days  Regularly busy   
6          Soda  ...  Mid-range product         Busiest days  Regularly busy   
7          Soda  ...  Mid-range product         Busiest days  Regularly busy   
8          Soda  ...  Mid-range product         Busiest days  Regularly busy   
9          Soda  ...  Mid-range product      Least busy days      Least b

In [43]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_aggregated.pkl'))